Objects/unionobject.c

```diff
@@ -137,102 +137,155 @@ union_richcompare(PyObject *a, PyObject *b, int op)
-            Py_ssize_t nested_arg_length = PyTuple_GET_SIZE(nested_args);
-            for (Py_ssize_t j = 0; j < nested_arg_length; j++) {
-                PyObject* nested_arg = PyTuple_GET_ITEM(nested_args, j);
-                Py_INCREF(nested_arg);
-                PyTuple_SET_ITEM(flattened_args, pos, nested_arg);
-                pos++;
+
+    for (int i = 0; i < args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(left_args, i);
+        Py_INCREF(arg);
+        PyTuple_SET_ITEM(tuple, i, arg);
+    }
+
+    Py_ssize_t pos = args_length;
+
+    for (int i = 0; i < other_args_length; i++) {
+        PyObject *arg = PyTuple_GET_ITEM(right_args, i);
+
+        int is_duplicate = 0;
+
+        for (int j = 0; j < args_length; j++) {
+            PyObject *left_arg = PyTuple_GET_ITEM(left_args, j);
+            is_duplicate = is_same(left_arg, arg);
+
+            if (is_duplicate < 0) {
+                Py_DECREF(tuple);
+                return NULL;
             }
-        } else {
-            if (arg == Py_None) {
-                arg = (PyObject *)&_PyNone_Type;
+            if (is_duplicate) {
+                break;
             }
+        }
+
+        if (!is_duplicate) {
             Py_INCREF(arg);
-            PyTuple_SET_ITEM(flattened_args, pos, arg);
+            PyTuple_SET_ITEM(tuple, pos, arg);
             pos++;
         }
     }
-    assert(pos == total_args);
-    return flattened_args;
-}
 
-static PyObject*
-dedup_and_flatten_args(PyObject* args)
-{
-    args = flatten_args(args);
-    if (args == NULL) {
+    if(_PyTuple_Resize(&tuple, pos) < 0)  {
+        Py_DECREF(tuple);
         return NULL;
     }
-    Py_ssize_t arg_length = PyTuple_GET_SIZE(args);
-    PyObject *new_args = PyTuple_New(arg_length);
-    if (new_args == NULL) {
-        Py_DECREF(args);
+
+    return tuple;
+}
+
+static PyObject*
+merge_union_and_obj(PyObject *left, PyObject *right) {
+    PyObject* args = ((unionobject *) left)->args;
+    Py_ssize_t args_length = PyTuple_GET_SIZE(args);
+
+    PyObject *tuple = PyTuple_New(args_length + 1);
+
+    if (tuple == NULL) {
         return NULL;
     }
-    // Add unique elements to an array.
-    Py_ssize_t added_items = 0;
-    for (Py_ssize_t i = 0; i < arg_length; i++) {
-        int is_duplicate = 0;
-        PyObject* i_element = PyTuple_GET_ITEM(args, i);
-        for (Py_ssize_t j = 0; j < added_items; j++) {
-            PyObject* j_element = PyTuple_GET_ITEM(new_args, j);
-            int is_ga = _PyGenericAlias_Check(i_element) &&
-                        _PyGenericAlias_Check(j_element);
-            // RichCompare to also deduplicate GenericAlias types (slower)
-            is_duplicate = is_ga ? PyObject_RichCompareBool(i_element, j_element, Py_EQ)
-                : i_element == j_element;
-            // Should only happen if RichCompare fails
+
+    int is_duplicate = 0;
+
+    for (int i = 0; i < args_length; i++) {
```

Comment:
This loop is similar to the loop in `merge_obj_and_union()` I think the code can be shared.

Pseudocode for `merge_union_and_obj()`:
```
if right in left.__args__:
    newargs = left.__args__
else:
    newargs = (*left.__args__, right)
```
Pseudocode for `merge_obj_and_union()`:
```
if left in right.__args__:
    newargs = right.__args__
else:
    newargs = (left, *right.__args__)
```
And you perhaps can use `PySequence_Contains()` or add a local specialized function if it is more efficient.

Reply:
That makes sense, will update the code. Thanks

Reply:
> Pseudocode for merge_obj_and_union()

This one isn't correct cause it won't match current behavior:
```py
int | (float | int)  -> int | float
```
But with proposed code it will be:
```py
int | (float | int)  -> float | int
```